src/Authentication.Abstractions/AzureSession.cs

```diff
@@ -236,24 +236,27 @@ public void RegisterComponent<T>(string componentName, Func<T> componentInitiali
             ChangeRegistry(
                 () =>
                 {
-                    var key = new ComponentKey(componentName, typeof(T));
-                    // todo: double check if dictionary operations are thread safe
-                    if (!_componentRegistry.ContainsKey(key) || overwrite)
+                    sessionLock.EnterWriteLock();
+                    try
                     {
-                        if (_componentRegistry.ContainsKey(key) && overwrite)
+                        var key = new ComponentKey(componentName, typeof(T));
+                        if (!_componentRegistry.TryGetValue(key, out var existingComponent) || overwrite)
                         {
-                            var existed = _componentRegistry[key];
-                            if (existed is IAzureSessionListener existedListener)
+                            if (overwrite && existingComponent is IAzureSessionListener existedListener)
                             {
                                 _eventHandler -= existedListener.OnEvent;
                             }
+                            var component = componentInitializer();
+                            _componentRegistry[key] = component;
```

Comment:
this line might also need atomic method applied

Reply:
Used the sessionLock to ensure thread safety.

Reply:
1. we are already using concurrentdictionary to achieve thread safe, I don't think it is proper to introduce another mechanism writelock here
2. if we are trying to use writelock, we do not need trygetvalue() either
3. if we are trying to mix atomic method from concurrentdictionary and writelock, the try/finally block should be applied on line 250 only

Reply:
not sure how atomic methods interact with locks, can you provide more insights?

Reply:
1. ReadWriteSlimLock has already been applied in this class. ConcurrentDictionary and lock has no conflicts. ConcurrentDictionary only makes sure the dictionary won't be updated at the same time by different threads (achieved by internal lock) while lock ensures only a single thread can execute the block of codes.
Furthermore, the _componentRegistry is defined as `IDictionary` (definition) instead of `ConcurrentDictionary` (implementation). So we are not able to invoke the TryAdd/TryUpdate/TryRemove without explict type cast. Maybe we can consider redefining its type to `ConcurrentDictionary`.
2. `TryGetValue` is highly recommended getting values. Contains + Get is legacy due to performance overhead especially for large objects. In all means, we should use TryGetValue to retrieve values.
3. I believe we should ensure only one thread can register/unregister the component instead of updating the components dictionary only.